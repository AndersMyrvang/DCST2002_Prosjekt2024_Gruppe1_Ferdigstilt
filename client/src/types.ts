// Frontend Types er samlet her

//Brukes i league-service
export type League = {
  id: number;
  name: string;
  country_id: number;
  country_name: string;
  emblem_image_url: string;
  page_id: number;
  content: string;
};

//Brukes i sideHistorikken
export type Page = {
  page_id?: number;
  created_by?: number;
  created_at: Date;
  view_count: number;
};

export type Revisions = {
  revision_id: number;
  content: string;
  revised_by: number;
  revised_at: Date;
  page_id: number;
  username?: string;
};

//Brukes i player-service
export type Player = {
  id: number;
  name: string;
  birth_date: Date;
  height: number;
  country_id: number;
  country_name: string;
  team_name: string;
  team_id: number;
  picture_url: string;
  page_id: number;
  content: string;
  league_id?: number;
  league_name?: string;
};

export type Country = {
  country_name: string;
  flag_image_url: string;
  country_id: number;
};

//Brukes ogs√• i team-service
export type Team = {
  name: string;
  id: number;
  country_id: number;
  country_name: string;
  coach: string;
  league_id: number;
  league_name: string;
  emblem_image_url: string;
  page_id: number;
  content: string;
};

//Brukes i comment-service
export type Comment = {
  comment_id: number;
  user_id: number;
  content: string;
  created_at: Date;
  updated_at: Date;
  page_id: number;
  likes: number;
  profile_image_url?: string;
  username?: string;
};

// Brukes i profile-component
export type UserProfile = {
  username: string;
  email: string;
  photo: string;
  firstLogin: string;
  lastLogin: string;
  is_admin: boolean;
  user_id: number;
};

// Brukes i searchBar-component
export type Suggestion = {
  type: string;
  id?: number;
  name: string;
};

// Brukes i API-respons for brukerdata (backend)
export type PublicUser = {
  username: string;
  email: string | null;
  photo: string | null;
  firstLogin: string;
  lastLogin: string;
  is_admin: boolean;
  user_id: number;
};

export type User = {
  user_id: number;
  google_id: string;
  username: string;
  email?: string | null;
  profile_image_url?: string | null;
  first_login: Date;
  last_login: Date;
  last_logout?: Date | null;
  is_admin: boolean;
};

export type Tag = {
  tag_id: number;
  tag_name: string;
};
