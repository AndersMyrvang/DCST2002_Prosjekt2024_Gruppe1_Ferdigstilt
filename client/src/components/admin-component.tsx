import React, { useState, useEffect } from 'react';
import { User } from '../types';

const AdminPanel = () => {
  const [users, setUsers] = useState<User[]>([]);

  const fetchUsers = () => {
    fetch('/api/users')
      .then((response) => response.json())
      .then((data) => setUsers(data));
  };

  useEffect(() => {
    fetchUsers();

    const interval = setInterval(fetchUsers, 60000);

    return () => clearInterval(interval);
  }, []);

  // Sjekker om en bruker er online basert på last_login og last_logout
  const isUserOnline = (user: User) => {
    if (
      user.last_login &&
      (!user.last_logout || new Date(user.last_login) > new Date(user.last_logout))
    ) {
      return true; 
    }
    return false;
  };

  // Funksjon for å endre admin-status
  const toggleAdminStatus = (userId: number, isAdmin: boolean) => {
    fetch(`/api/users/${userId}/set-admin`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ isAdmin: !isAdmin }),
    })
      .then((response) => {
        if (response.ok) {
          setUsers(
            users.map((user) => (user.user_id === userId ? { ...user, is_admin: !isAdmin } : user)),
          );
        } else {
          console.error('Feil ved oppdatering av admin-status');
        }
      })
      .catch((error) => console.error('Feil ved toggleAdminStatus:', error));
  };

  return (
    <div className="admin-panel">
      <table>
        <thead>
          <tr>
            <th>User ID</th>
            <th>Username</th>
            <th>Email</th>
            <th>Member Since</th>
            <th>Online Status</th>
            <th>Admin</th>
          </tr>
        </thead>
        <tbody>
          {users.map((user) => (
            <tr key={user.user_id}>
              <td>{user.user_id}</td>
              <td>{user.username}</td>
              <td>{user.email}</td>
              <td>{new Date(user.first_login).toLocaleDateString()}</td>
              <td>
                <span className={isUserOnline(user) ? 'online-dot' : 'offline-dot'}></span>
              </td>
              <td>
                <input
                  type="checkbox"
                  checked={user.is_admin}
                  onChange={() => toggleAdminStatus(user.user_id!, user.is_admin)}
                />
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default AdminPanel;
