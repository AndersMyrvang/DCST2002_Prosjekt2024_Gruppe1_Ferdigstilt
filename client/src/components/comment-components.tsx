import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Row, Column, Form, Button, Card2, Card3 } from '../widgets';
import { NavLink } from 'react-router-dom';
import commentService from '../services/comment-service';
import userService from '../services/user-service';
import { Comment, User } from '../types';
import { createHashHistory } from 'history';
import { AiFillLike, AiFillDislike } from 'react-icons/ai';
import { FaTrashAlt } from 'react-icons/fa';
import { BiSolidPencil } from 'react-icons/bi';

const history = createHashHistory();

interface CommentSectionProps {
  page_id: number;
  onNewComment: () => void;
}

//Komponent for ny kommentar
export class NewComment extends Component<CommentSectionProps> {
  comment: Comment = {
    comment_id: 0,
    user_id: 0,
    content: '',
    created_at: new Date(),
    updated_at: new Date(),
    page_id: 0,
    likes: 0,
  };

  render() {
    return (
      <>
        <Card title="New Comment">
          <Column>
            <Form.Input
              type="text"
              aria-label={`Comment input`}
              value={this.comment.content}
              onChange={(event) => (this.comment.content = event.currentTarget.value)}
              rows={3}
            />
          </Column>
          <Column>
            <Button.Success
              onClick={() => {
                this.NewComment(this.comment.user_id, this.comment.content, this.props.page_id);
                this.comment.content = '';
              }}
            >
              {' '}
              SEND{' '}
            </Button.Success>
          </Column>
        </Card>
      </>
    );
  }
  mounted() {
    const currentUserId = localStorage.getItem('user_id');
    if (currentUserId != null) this.comment.user_id = parseInt(currentUserId, 10);
  }
  NewComment(user_id: number, content: string, page_id: number) {
    commentService
      .create(user_id, content, page_id)
      .then((comment) => {
        this.props.onNewComment();
      })
      .catch((error) => {
        console.error('Error creating comment:', error);
      });
  }
}

//Komponent for kommentarseksjonen (visning av kommentarer p√• en spesifikk side)
export class CommentSection extends Component<CommentSectionProps> {
  comment: Comment[] = [];
  user: User = {
    user_id: 0,
    google_id: '',
    username: '',
    email: '',
    profile_image_url: '',
    first_login: new Date(),
    last_login: new Date(),
    last_logout: new Date(),
    is_admin: false,
  };
  updatingComment: number | null = null;
  updatedContent: string = '';

  render() {
    return (
      <>
        <Card title="Comments">
          {this.user.user_id != 0 ? (
            <NewComment page_id={this.props.page_id} onNewComment={() => this.mounted()} />
          ) : null}
          {this.comment.map((comment) => (
            <Row key={new Date(comment.created_at).getTime()}>
              <Column>
                {' '}
                <img
                  src={`${comment.profile_image_url}`}
                  alt={`Bilde fra ${comment.profile_image_url}`}
                  width={'45px'}
                  style={{
                    borderRadius: '50%',
                    objectFit: 'cover',
                  }}
                />
              </Column>
              <Column>
                <b>{comment.username}</b>
              </Column>
              <Column>
                {this.updatingComment == comment.comment_id ? (
                  <Form.Input
                    type="text"
                    value={this.updatedContent}
                    onChange={(event) => (this.updatedContent = event.currentTarget.value)}
                  />
                ) : (
                  comment.content
                )}
              </Column>

              <Column>
                {this.user.user_id == comment.user_id || this.user.is_admin == true ? (
                  <Button.Danger
                    onClick={() => {
                      this.deleteComment(comment.comment_id);
                      this.mounted();
                    }}
                  >
                    <FaTrashAlt />
                  </Button.Danger>
                ) : null}
              </Column>
              <Column>
                {this.user.user_id == comment.user_id || this.user.is_admin == true ? (
                  <>
                    {this.updatingComment == comment.comment_id ? (
                      <>
                        <Button.Success
                          onClick={() => {
                            this.updateComment(comment.comment_id, this.updatedContent);
                            this.updatingComment = null;
                            this.mounted();
                          }}
                        >
                          Lagre
                        </Button.Success>
                      </>
                    ) : (
                      <>
                        <Button.Light
                          onClick={() => {
                            this.updatingComment = comment.comment_id;
                            this.updatedContent = comment.content;
                          }}
                        >
                          <BiSolidPencil />
                        </Button.Light>
                      </>
                    )}
                  </>
                ) : null}
              </Column>
            </Row>
          ))}
        </Card>
      </>
    );
  }

  mounted() {
    const current_userId = localStorage.getItem('user_id');
    if (current_userId != null) {
      userService
        .getUserById(parseInt(current_userId, 10))
        .then((user: User) => {
          this.user = user;
        })
        .catch((error) => {
          console.error('Error getting user: ', error);
          Alert.danger('Error getting user: ' + error.message);
        });
    }

    commentService
      .listComments(this.props.page_id)
      .then(async (comments: Comment[]) => {
        this.comment = comments;
      })
      .catch((error) => {
        console.error('Error fetching comments:', error);
        Alert.danger('Error getting comments: ' + error.message);
      });
  }

  deleteComment(comment_id: number) {
    commentService.deleteComment(comment_id);
  }
  updateComment(comment_id: number, content: string) {
    commentService.update(comment_id, content);
  }
}

export default CommentSection;
