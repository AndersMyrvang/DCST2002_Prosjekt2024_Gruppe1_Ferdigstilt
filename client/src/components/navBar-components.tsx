import React, { useState, useRef, useEffect } from 'react';
import SearchBar from './searchBar-component';
import { FaHouse } from 'react-icons/fa6';
import { User } from '../types';

interface HamburgerProps {
  user: User | null;
}

function DarkModeButton() {
  const [darkMode, setDarkMode] = useState(() => {
    const savedMode = localStorage.getItem('darkMode');
    return savedMode === 'true';
  });

  const toggleDarkMode = () => {
    const newMode = !darkMode;
    setDarkMode(newMode);
    document.body.classList.toggle('dark-mode', newMode);
    localStorage.setItem('darkMode', newMode.toString());
  };

  useEffect(() => {
    if (darkMode) {
      document.body.classList.add('dark-mode');
    }
  }, [darkMode]);

  return (
    <button className="dark-mode-button" onClick={toggleDarkMode} aria-label="Toggle Dark Mode">
      <i
        className={darkMode ? 'fas fa-sun' : 'fas fa-moon'}
        style={{ fontSize: '24px', color: darkMode ? '#FFD700' : '#555' }}
      />
    </button>
  );
}

const Hamburger: React.FC<HamburgerProps> = ({ user }) => {
  const [burgerClass, setBurgerClass] = useState('burger-bar unclicked');
  const [menuClass, setMenuClass] = useState('menu hidden');
  const [isMenuClicked, setIsMenuClicked] = useState(false);

  const [isLoggedIn, setIsLoggedIn] = useState(() => localStorage.getItem('isLoggedIn') === 'true');

  const [userName, setUserName] = useState(localStorage.getItem('user') || '');
  const menuRef = useRef<HTMLDivElement>(null);

  const updateMenu = () => {
    setIsMenuClicked((prev) => !prev);
    setBurgerClass(isMenuClicked ? 'burger-bar unclicked' : 'burger-bar clicked');
    setMenuClass(isMenuClicked ? 'menu hidden' : 'menu visible');
  };

  const handleClickOutside = (event: MouseEvent) => {
    if (menuRef.current && !menuRef.current.contains(event.target as Node)) {
      closeMenu();
    }
  };

  const closeMenu = () => {
    setBurgerClass('burger-bar unclicked');
    setMenuClass('menu hidden');
    setIsMenuClicked(false);
  };

  const handleSearch = (query: string) => {
    fetch(`/api/search?q=${query}`)
      .then((response) => response.json())
      .then((data) => {})
      .catch((error) => {
        console.error('Feil ved søk:', error);
      });
  };

  const handleLogout = () => {
    fetch('/api/logout', { method: 'GET', credentials: 'include' })
      .then((response) => {
        if (response.ok) {
          setIsLoggedIn(false);
          setUserName('');
          localStorage.removeItem('isLoggedIn');
          localStorage.removeItem('user');
          localStorage.removeItem('user_id');
          window.location.href = '/';
        } else {
          console.error('Logout failed');
        }
      })
      .catch((error) => console.error('Logout error:', error));
  };

  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside);

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, []);

  // Sjekker om bruker er pålogget
  useEffect(() => {
    fetch('/api/current_user')
      .then((response) => response.json())
      .then((data) => {
        if (data && data.username) {
          setIsLoggedIn(true);
          setUserName(data.username);
          localStorage.setItem('isLoggedIn', 'true');
          localStorage.setItem('user', data.username);
          localStorage.setItem('user_id', data.user_id);
        } else {
          setIsLoggedIn(false);
          localStorage.setItem('isLoggedIn', 'false');
        }
      })
      .catch(() => {
        setIsLoggedIn(false);
        localStorage.setItem('isLoggedIn', 'false');
      });
  }, []);

  return (
    <div style={{ width: '100%' }} ref={menuRef}>
      <nav>
        <div className="house">
          <a href="/">
            <FaHouse />
          </a>
        </div>
        <div className="burger-menu" onClick={updateMenu} aria-label="Toggle Menu">
          <div className={burgerClass}></div>
          <div className={burgerClass}></div>
          <div className={burgerClass}></div>
        </div>

        <SearchBar onSearch={handleSearch} />

        <DarkModeButton />
      </nav>

      <div className={menuClass}>
        <ul className="menu-links" onClick={closeMenu} role="menu">
          <li>
            <a href="/">Home</a>
          </li>
          <li>
            <a href="#players">Players</a>
          </li>
          <li>
            <a href="#teams">Teams</a>
          </li>
          <li>
            <a href="#leagues">Leagues</a>
          </li>
          <li>
            <a href="#tags">Tags</a>
          </li>

          {!isLoggedIn ? (
            <li>
              <a href="https://accounts.google.com/signup/v2/createaccount">Register </a> /
              <a href="/auth/google"> Log in with Google </a>
            </li>
          ) : (
            <>
              <li>
                <a href="#/profile">My profile </a> /{' '}
                <a href="#" onClick={handleLogout}>
                  Log out
                </a>
              </li>
            </>
          )}
        </ul>
      </div>
    </div>
  );
};

export default Hamburger;
