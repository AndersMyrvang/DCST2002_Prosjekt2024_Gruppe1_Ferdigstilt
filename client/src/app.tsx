import React, { useEffect, useState } from 'react';
import { HashRouter as Router, Route } from 'react-router-dom';
import NavBar from './components/navBar-components';
import { PlayerDetails } from './components/player-components';
import { TeamDetails } from './components/team-components';
import { LeagueDetails } from './components/league-components';

// Dummy Home component as placeholder
function Home() {
  return <div>Welcome Home</div>;
}

function App() {
  interface User {
    displayName: string;
  }

  const [user, setUser] = useState<User | null>(null);

  // Fetching user info from /api/current_user when the app loads
  useEffect(() => {
    fetch('/api/current_user')
      .then((res) => res.json())
      .then((data) => {
        setUser(data);
      });
  }, []);

  return (
    <Router>
      {/* Define routes for the page */}
      <Route exact path="/" component={Home} />
      <Route path="/players/:name" component={PlayerDetails} />
      <Route path="/teams/:name" component={TeamDetails} />
      <Route path="/leagues/:name" component={LeagueDetails} />
    </Router>
  );
}

export default App;
