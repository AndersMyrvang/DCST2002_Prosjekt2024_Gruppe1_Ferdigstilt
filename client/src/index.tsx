import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert } from './widgets';
import {
  PlayerList,
  PlayerDetails,
  PlayerEdit,
  PlayerListAll,
  PlayerNew,
} from './components/player-components';
import {
  TeamList,
  TeamDetails,
  TeamEdit,
  TeamListAll,
  TeamNew,
} from './components/team-components';
import { TagList, TagAdd } from './components/tag-component';
import {
  LeagueList,
  LeagueDetails,
  LeagueEdit,
  LeagueListAll,
  LeagueNew,
} from './components/league-components';
import Hamburger from './components/navBar-components';
import Profile from './components/profile-component';

class Menu extends Component {
  render() {
    return <Hamburger user={null} />;
  }
}

class Home extends Component {
  render() {
    return (
      <Card title="Welcome">
        <PlayerList />
        <TeamList />
        <LeagueList />
      </Card>
    );
  }
}

let root = document.getElementById('root');
if (root)
  createRoot(root).render(
    <HashRouter>
      <div>
        <Alert />
        <Menu />
        <Route exact path="/" component={Home} />
        <Route exact path="/menu" component={Hamburger} />

        <Route exact path="/page" component={PlayerList} />
        <Route exact path="/players" component={PlayerListAll} />
        <Route exact path="/player/:id" component={PlayerDetails} />
        <Route exact path="/player/:id/edit" component={PlayerEdit} />
        <Route exact path="/player_id/new" component={PlayerNew} />

        <Route exact path="/page" component={TeamList} />
        <Route exact path="/teams" component={TeamListAll} />
        <Route exact path="/team/:id/" component={TeamDetails} />
        <Route exact path="/team/edit/:id/" component={TeamEdit} />
        <Route exact path="/team_id/new" component={TeamNew} />

        <Route exact path="/page" component={LeagueList} />
        <Route exact path="/leagues" component={LeagueListAll} />
        <Route exact path="/league/:id" component={LeagueDetails} />
        <Route exact path="/league/:id/edit" component={LeagueEdit} />
        <Route exact path="/league_id/new" component={LeagueNew} />

        <Route exact path="/tags" component={TagList} />
        <Route exact path="/tags/new" component={TagAdd} />
        <Route exact path="/profile" component={Profile} />
      </div>
    </HashRouter>,
  );
