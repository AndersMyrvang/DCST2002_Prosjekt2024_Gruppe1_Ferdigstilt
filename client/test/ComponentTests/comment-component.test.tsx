import * as React from 'react';
import { shallow } from 'enzyme';
import { CommentSection, NewComment } from '../../src/components/comment-components';
import { Form, Button, Column } from '../../src/widgets';
import { NavLink } from 'react-router-dom';
import commentService from '../../src/services/comment-service';
import userService from '../../src/services/user-service';
import { Comment, User } from '../../src/types';
import { FaTrashAlt } from 'react-icons/fa';
import { BiSolidPencil } from 'react-icons/bi';

jest.mock('../../src/services/comment-service', () => ({
  create: jest.fn(() => Promise.resolve({ comment_id: 4 })),
  listComments: jest.fn(() =>
    Promise.resolve([
      { comment_id: 1, user_id: 1, content: 'Great post!', likes: 5 },
      { comment_id: 2, user_id: 2, content: 'Nice article!', likes: 3 },
    ]),
  ),
  deleteComment: jest.fn(() => Promise.resolve()),
  update: jest.fn(() =>
    Promise.resolve({
      comment: { comment_id: 1, user_id: 1, content: 'Updated comment text', likes: 5 },
    }),
  ),
}));

jest.mock('../../src/services/user-service', () => ({
  getUserById: jest.fn((user_id) =>
    Promise.resolve({
      user_id,
      username: 'testuser',
      is_admin: false,
    }),
  ),
}));

describe('Comment component tests', () => {
  const mockOnNewComment = jest.fn();
  const page_id = 1;

  beforeEach(() => {
    jest.spyOn(window.localStorage.__proto__, 'getItem').mockImplementation((key) => {
      if (key === 'user_id') return '1'; 
      return null;
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  test('NewComment component renders and posts comment', async () => {
    const wrapper = shallow(<NewComment page_id={page_id} onNewComment={mockOnNewComment} />);

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'New comment text' } });
    wrapper.find(Button.Success).simulate('click');

    await new Promise((resolve) => setTimeout(resolve, 0));

    expect(commentService.create).toHaveBeenCalledWith(1, 'New comment text', page_id);
    expect(mockOnNewComment).toHaveBeenCalled();
  });

  test('CommentSection loads and displays comments', async () => {
    const wrapper = shallow(<CommentSection page_id={page_id} onNewComment={mockOnNewComment} />);

    await new Promise((resolve) => setTimeout(resolve, 0));

    expect(commentService.listComments).toHaveBeenCalledWith(page_id);
    expect(wrapper.containsMatchingElement(<Column>Great post!</Column>)).toEqual(true);
    expect(wrapper.containsMatchingElement(<Column>Nice article!</Column>)).toEqual(true);
  });

  test('CommentSection deletes a comment and reloads', async () => {
    const wrapper = shallow(<CommentSection page_id={page_id} onNewComment={mockOnNewComment} />);

    await new Promise((resolve) => setTimeout(resolve, 0));

    wrapper.setState({ user: { user_id: 1, username: 'testuser', is_admin: true } });

    wrapper.find(Button.Danger).at(0).simulate('click');
    await new Promise((resolve) => setTimeout(resolve, 0));

    expect(commentService.deleteComment).toHaveBeenCalledWith(1);
    expect(commentService.listComments).toHaveBeenCalledWith(page_id);
  });

  test('NewComment snapshot matches', () => {
    const wrapper = shallow(<NewComment page_id={page_id} onNewComment={mockOnNewComment} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('CommentSection snapshot matches', async () => {
    const wrapper = shallow(<CommentSection page_id={page_id} onNewComment={mockOnNewComment} />);
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(wrapper).toMatchSnapshot();
  });

  test('NewComment renders correctly initially', () => {
    const wrapper = shallow(<NewComment page_id={page_id} onNewComment={mockOnNewComment} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('CommentSection renders correctly with loaded comments', async () => {
    const wrapper = shallow(<CommentSection page_id={page_id} onNewComment={mockOnNewComment} />);

    await new Promise((resolve) => setTimeout(resolve, 0));
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  test('CommentSection renders correctly when editing a comment', async () => {
    const wrapper = shallow(<CommentSection page_id={page_id} onNewComment={mockOnNewComment} />);

    await new Promise((resolve) => setTimeout(resolve, 0));
    wrapper.setState({ updatingComment: 1, updatedContent: 'Test edit content' });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  test("CommentSection renders correctly with delete and edit buttons for user's own comment", async () => {
    const wrapper = shallow(<CommentSection page_id={page_id} onNewComment={mockOnNewComment} />);

    wrapper.setState({
      user: { user_id: 1, is_admin: false },
      comment: [{ comment_id: 1, user_id: 1, content: "User's own comment", likes: 0 }],
    });

    await new Promise((resolve) => setTimeout(resolve, 0));
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  test('CommentSection renders error alert when comments fail to load', async () => {
    jest
      .spyOn(commentService, 'listComments')
      .mockRejectedValue(new Error('Error loading comments'));

    const wrapper = shallow(<CommentSection page_id={page_id} onNewComment={mockOnNewComment} />);

    await new Promise((resolve) => setTimeout(resolve, 0));
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });
});
