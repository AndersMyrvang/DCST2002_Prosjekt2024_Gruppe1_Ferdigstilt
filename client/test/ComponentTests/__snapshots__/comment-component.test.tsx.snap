// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment component tests CommentSection renders correctly when editing a comment 1`] = `
<Fragment>
  <Card
    title="Comments"
  >
    <NewComment
      onNewComment={[Function]}
      page_id={1}
    />
    <Row
      key="NaN"
    >
      <Column>
         
        <img
          alt="Bilde fra undefined"
          src="undefined"
          style={
            {
              "borderRadius": "50%",
              "objectFit": "cover",
            }
          }
          width="45px"
        />
      </Column>
      <Column>
        <b />
      </Column>
      <Column>
        Great post!
      </Column>
      <Column>
        <ButtonDanger
          onClick={[Function]}
        >
          <FaTrashAlt />
        </ButtonDanger>
      </Column>
      <Column>
        <ButtonLight
          onClick={[Function]}
        >
          <BiSolidPencil />
        </ButtonLight>
      </Column>
    </Row>
    <Row
      key="NaN"
    >
      <Column>
         
        <img
          alt="Bilde fra undefined"
          src="undefined"
          style={
            {
              "borderRadius": "50%",
              "objectFit": "cover",
            }
          }
          width="45px"
        />
      </Column>
      <Column>
        <b />
      </Column>
      <Column>
        Nice article!
      </Column>
      <Column />
      <Column />
    </Row>
  </Card>
</Fragment>
`;

exports[`Comment component tests CommentSection renders correctly with delete and edit buttons for user's own comment 1`] = `
<Fragment>
  <Card
    title="Comments"
  >
    <NewComment
      onNewComment={[Function]}
      page_id={1}
    />
    <Row
      key="NaN"
    >
      <Column>
         
        <img
          alt="Bilde fra undefined"
          src="undefined"
          style={
            {
              "borderRadius": "50%",
              "objectFit": "cover",
            }
          }
          width="45px"
        />
      </Column>
      <Column>
        <b />
      </Column>
      <Column>
        Great post!
      </Column>
      <Column>
        <ButtonDanger
          onClick={[Function]}
        >
          <FaTrashAlt />
        </ButtonDanger>
      </Column>
      <Column>
        <ButtonLight
          onClick={[Function]}
        >
          <BiSolidPencil />
        </ButtonLight>
      </Column>
    </Row>
    <Row
      key="NaN"
    >
      <Column>
         
        <img
          alt="Bilde fra undefined"
          src="undefined"
          style={
            {
              "borderRadius": "50%",
              "objectFit": "cover",
            }
          }
          width="45px"
        />
      </Column>
      <Column>
        <b />
      </Column>
      <Column>
        Nice article!
      </Column>
      <Column />
      <Column />
    </Row>
  </Card>
</Fragment>
`;

exports[`Comment component tests CommentSection renders correctly with loaded comments 1`] = `
<Fragment>
  <Card
    title="Comments"
  >
    <NewComment
      onNewComment={[Function]}
      page_id={1}
    />
    <Row
      key="NaN"
    >
      <Column>
         
        <img
          alt="Bilde fra undefined"
          src="undefined"
          style={
            {
              "borderRadius": "50%",
              "objectFit": "cover",
            }
          }
          width="45px"
        />
      </Column>
      <Column>
        <b />
      </Column>
      <Column>
        Great post!
      </Column>
      <Column>
        <ButtonDanger
          onClick={[Function]}
        >
          <FaTrashAlt />
        </ButtonDanger>
      </Column>
      <Column>
        <ButtonLight
          onClick={[Function]}
        >
          <BiSolidPencil />
        </ButtonLight>
      </Column>
    </Row>
    <Row
      key="NaN"
    >
      <Column>
         
        <img
          alt="Bilde fra undefined"
          src="undefined"
          style={
            {
              "borderRadius": "50%",
              "objectFit": "cover",
            }
          }
          width="45px"
        />
      </Column>
      <Column>
        <b />
      </Column>
      <Column>
        Nice article!
      </Column>
      <Column />
      <Column />
    </Row>
  </Card>
</Fragment>
`;

exports[`Comment component tests CommentSection renders error alert when comments fail to load 1`] = `
<Fragment>
  <Card
    title="Comments"
  >
    <NewComment
      onNewComment={[Function]}
      page_id={1}
    />
  </Card>
</Fragment>
`;

exports[`Comment component tests CommentSection snapshot matches 1`] = `
<Fragment>
  <Card
    title="Comments"
  >
    <NewComment
      onNewComment={[Function]}
      page_id={1}
    />
    <Row
      key="NaN"
    >
      <Column>
         
        <img
          alt="Bilde fra undefined"
          src="undefined"
          style={
            {
              "borderRadius": "50%",
              "objectFit": "cover",
            }
          }
          width="45px"
        />
      </Column>
      <Column>
        <b />
      </Column>
      <Column>
        Great post!
      </Column>
      <Column>
        <ButtonDanger
          onClick={[Function]}
        >
          <FaTrashAlt />
        </ButtonDanger>
      </Column>
      <Column>
        <ButtonLight
          onClick={[Function]}
        >
          <BiSolidPencil />
        </ButtonLight>
      </Column>
    </Row>
    <Row
      key="NaN"
    >
      <Column>
         
        <img
          alt="Bilde fra undefined"
          src="undefined"
          style={
            {
              "borderRadius": "50%",
              "objectFit": "cover",
            }
          }
          width="45px"
        />
      </Column>
      <Column>
        <b />
      </Column>
      <Column>
        Nice article!
      </Column>
      <Column />
      <Column />
    </Row>
  </Card>
</Fragment>
`;

exports[`Comment component tests NewComment renders correctly initially 1`] = `
<Fragment>
  <Card
    title="New Comment"
  >
    <Column>
      <FormInput
        aria-label="Comment input"
        onChange={[Function]}
        rows={3}
        type="text"
        value=""
      />
    </Column>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
         
        SEND
         
      </ButtonSuccess>
    </Column>
  </Card>
</Fragment>
`;

exports[`Comment component tests NewComment snapshot matches 1`] = `
<Fragment>
  <Card
    title="New Comment"
  >
    <Column>
      <FormInput
        aria-label="Comment input"
        onChange={[Function]}
        rows={3}
        type="text"
        value=""
      />
    </Column>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
         
        SEND
         
      </ButtonSuccess>
    </Column>
  </Card>
</Fragment>
`;
